ctsit/nacculator
# Add optional form logic to FTLD, update documentation, generalize redcap_event_names

This branch introduces several smaller maintenance-level updates. 
The FTLD builder files now function similar to the UDS3 module builders, where NACCulator will first scan the Z1X to see which forms are present in the record rather than automatically trying to process all possible forms. This will prevent users from having to repeatedly manually delete empty optional forms that will never be filled out. 
NACCulator's event detection has been generalized to be more compatible with more centers. The requirement for the initial visit packet's event name to contain "initial_visit" has been changed to "initial", and the requirement for the followup visit packet's event event name to contain "followup_visit" has been changed to just "follow" (to account for variants with either "followup" OR "follow_up").
The README has been updated to include more information about what data NACCulator needs from REDCap in order to run correctly (such as the filled Z1X form), and updated documentation on running the filters. I corrected the example commands for running the individual filters (with an input csv AND an output csv) so that they work properly now. All filter commands now require an argument specifying the config file so that the "validate" function in filters.py works properly. This makes the command structure consistent across all filters.

This change was made to reflect feedback from users at Houston Methodist and Harvard. Ideally, it will make NACCulator more simple and straightforward to use.


## Verification

_List the steps needed to make sure this thing works. Be precise._

To test filters:
1. Make sure you have your `nacculator_config.cfg` file prepared.
2. Collect a `current-db-subjects.csv` file.
3. Add the path to `current-db-subjects.csv` to the config file.
4. From the command-line, run:
`redcap2nacc -f cleanPtid -meta nacculator_cfg.ini <data_input.csv >filtered_output.csv
4a. Run any other filters besides cleanPtid with this same format if desired.
5. The output should look like this: a file called `filtered_output.csv` that has data with the specified filter applied to it.

To test FTLD optional forms:
1. Collect a REDCap export csv of sample FTLD data (`data_ftld.csv` is on jira; file type not accepted on github).
2. From the command-line, run:
`redcap2nacc -ftld -ivp <data_ftld.csv >data_ftld.txt`
3. The output should look like this: a regular NACCulator output .txt with FTLD module data.
4. If you pass data with any errors, the program will skip that record and print the error.
5. If you pass data with the wrong redcap_event_name or an unfilled Z1X, NACCulator will skip that record without printing any output.



- Verify the thing does this: Converts a .csv file either to a NACC-ready output .txt or another filtered .csv file.
- Verify the thing does not do that: Raises errors when there are none, and doesn't raise errors when they are present.


## Affirmations

All of these should have a check by them. Any exception requires an explanation.

* [x] I tested the code with every flag option.
